web3/middleware/filter.py

```diff
@@ -0,0 +1,324 @@
+    (0, 3)
+    >>> next(segment_counter)
+    (3, 6)
+    >>> next(segment_counter)
+    (6, 9)
+    >>> next(segment_counter) #  Remainder is also returned
+    (9, 10)
+    """
+    def finite(start, stop, step):
+        #  If the initial range is less than the step
+        #  just return (start, stop)
+        if stop < step:
+            yield (start, stop)
+            return
+        for i in zip(
+                range(start, stop - step, step),
+                range(start + step, stop, step)):
+            yield i
+        else:
+            remainder = (stop - start) % step
+            #  Handle the remainder
+            if remainder:
+                yield (stop - remainder, stop)
+
+    def infinite(start, step):
+        for i in zip(
+                count(start, step),
+                count(start + step, step)):
+            yield i
+
+    if stop is None:
+        return infinite(start, step)
+    else:
+        return finite(start, stop, step)
+
+
+def block_range_chunker(start_block, last_block=None, step=5):
+    """Returns 2-tuple ranges describing chunks of block from start_block to last_block
+
+       Segments do not overlap to facilitate use as ``toBlock``, ``fromBlock``
+       json-rpc arguments, which are both inclusive.
+    """
+
+    if last_block is not None and start_block > last_block:
+        raise TypeError(
+            "Incompatible start and stop arguments.",
+            "Start must be less than or equal to stop.")
+
+    def block_range_chunker(start_block, last_block, step):
+        for from_block, to_block in segment_count(start_block, last_block + 1, step):
+            yield (from_block, to_block - 1)
+
+    return block_range_chunker(start_block, last_block, step)
+
+
+def iter_latest_block(web3, to_block=None):
+    """Returns a generator that dispenses the latest block, if
+    any new blocks have been mined since last iteration.
+
+    If there are no new blocks None is returned.
+
+    If ``to_block`` is defined, ``StopIteration`` is raised
+    after to_block is reached.
+
+    >>> mined_blocks = dispense_mined_blocks(web3, 0, 10)
+    >>> next(new_blocks)  # Latest block = 0
+    0
+    >>> next(new_blocks)  # No new blocks
+    >>> next(new_blocks)  # Latest block = 1
+    1
+    >>> next(new_blocks)  # Latest block = 10
+    10
+    >>> next(new_blocks)
+    Traceback (most recent call last):
+      File "<stdin>", line 1, in <module>
+      StopIteration
+    >>>
+    """
+    _last = None
+    while True:
+        latest_block = web3.eth.blockNumber
+        if (latest_block is not None and
+                to_block is not None and
+                latest_block > to_block):
+            return
+        #  No new blocks since last iteration.
+        if _last is not None and _last == latest_block:
+            yield None
+        else:
+            yield latest_block
+        _last = latest_block
+
+
+def count(start=0, step=1):
+    """A counter you can update with send()
```

Comment:
This seems like a code example in the docstring would be very informative because even reading this it's not entirely clear how I would go about using it.